### PR TITLE
[LLVM 4.0] Fix CreateCompileUnit

### PR DESCRIPTION
**Pull request by mattico**
This is largely identical to @dylanmckay's [patch](https://github.com/dylanmckay), except that it doesn't try to use `file_metadata()`. I don't think that is necessary because we don't want the compile unit to be added to  `debug_context.created_files`, though I'd like confirmation from someone who knows for sure. If that is needed, I can modify `file_metadata_()` so that it can be used from `compile_unit_metadata()`.

### CODE REVIEW

**Comment by mattico** [comment]
cc #37609

File: src/librustc_llvm/ffi.rs [hunk 1]
```diff
@@ -1334,8 +1334,7 @@ extern "C" {
 
     pub fn LLVMRustDIBuilderCreateCompileUnit(Builder: DIBuilderRef,
                                               Lang: c_uint,
-                                              File: *const c_char,
-                                              Dir: *const c_char,
+                                              File: DIFile,
```
Comment (dylanmckay) [inline src/librustc_llvm/ffi.rs]: This looks really nice!

File: src/librustc_trans/debuginfo/metadata.rs [hunk 2]
```diff
@@ -787,17 +787,20 @@ pub fn compile_unit_metadata(scc: &SharedCrateContext,
                            (option_env!("CFG_VERSION")).expect("CFG_VERSION"));
 
     let compile_unit_name = compile_unit_name.as_ptr();
-    let work_dir = path2cstr(&work_dir);
-    let producer = CString::new(producer).unwrap();
+    let work_dir = path2cstr(&work_dir).as_ptr();
+    let producer = CString::new(producer).unwrap().as_ptr();
```
Comment (rkruppe) [inline src/librustc_trans/debuginfo/metadata.rs]: This line is wrong: it will drop the `CString` and keep a raw pointer to its now-freed buffer around.

Comment (mattico) [inline src/librustc_trans/debuginfo/metadata.rs]: Ah, I was wondering why the `.as_ptr()` calls didn't match the line above.

**Pull request merged**